File: agent-python/tools/registry.py
"""
工具注册表 - 管理所有工具实例
"""

import json
from typing import Dict, List, Any
from .base import BaseTool

class ToolRegistry:
    """工具注册表"""
    
    def __init__(self):
        self.tools: Dict[str, BaseTool] = {}
    
    def register(self, tool: BaseTool) -> None:
        """注册工具"""
        name = tool.get_name()
        self.tools[name] = tool
        print(f"已注册工具: {name}")
    
    def get_tool(self, name: str) -> BaseTool:
        """获取工具"""
        if name not in self.tools:
            raise ValueError(f"工具 '{name}' 未找到")
        return self.tools[name]
    
    def get_functions_xml(self) -> str:
        """生成 <functions> XML 块，用于 LLM 上下文"""
        functions_list = []
        
        for tool in self.tools.values():
            schema = tool.get_function_schema()
            # 转换为紧凑的 JSON 格式
            json_str = json.dumps(schema, ensure_ascii=False, separators=(',', ':'))
            functions_list.append(f"<function>{json_str}</function>")
        
        return f"<functions>\n{chr(10).join(functions_list)}\n</functions>"
    
    def parse_function_calls(self, response: str) -> List[Dict[str, Any]]:
        """解析 LLM 响应中的 <function_calls> 块"""
        import re
        
        # 提取 function_calls 块
        function_calls_match = re.search(r'<function_calls>(.*?)</function_calls>', response, re.DOTALL)
        if not function_calls_match:
            return []
        
        function_calls_content = function_calls_match.group(1)
        
        # 提取所有 invoke 块
        invoke_matches = re.findall(r'<invoke name="([^"]+)">(.*?)</invoke>', function_calls_content, re.DOTALL)
        
        calls = []
        for tool_name, params_content in invoke_matches:
            # 解析参数
            param_matches = re.findall(r'<parameter name="([^"]+)">(.*?)</parameter>', params_content, re.DOTALL)
            parameters = {name: value.strip() for name, value in param_matches}
            
            calls.append({
                "tool_name": tool_name,
                "parameters": parameters
            })
        
        return calls
    
    def execute_function_calls(self, calls: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """执行工具调用列表"""
        results = []
        
        for call in calls:
            tool_name = call["tool_name"]
            parameters = call["parameters"]
            
            try:
                tool = self.get_tool(tool_name)
                result = tool.execute(parameters)
                results.append({
                    "tool_name": tool_name,
                    "success": True,
                    "result": result
                })
            except Exception as e:
                results.append({
                    "tool_name": tool_name,
                    "success": False,
                    "error": str(e)
                })
        
        return results 

File: agent-js/src/tools/registry.js
import { isValidTool } from './base.js';

export class ToolRegistry {
  constructor() {
    this.tools = new Map();
  }

  /**
   * 注册工具到系统
   * @param {Object} tool
   */
  register(tool) {
    if (!isValidTool(tool)) {
      throw new Error('Invalid tool: must implement required methods');
    }
    
    const name = tool.getName();
    this.tools.set(name, tool);
    console.log(`Tool registered: ${name}`);
  }

  /**
   * 获取工具实例
   * @param {string} name
   * @returns {Object|null}
   */
  getTool(name) {
    return this.tools.get(name) || null;
  }

  /**
   * 获取所有工具名称
   * @returns {string[]}
   */
  getToolNames() {
    return Array.from(this.tools.keys());
  }

  /**
   * 生成标准 <functions> XML 块
   * @returns {string}
   */
  getFunctionsXML() {
    if (this.tools.size === 0) {
      return '';
    }

    const functions = Array.from(this.tools.values())
      .map(tool => {
        const schema = tool.getFunctionSchema();
        return `<function>${JSON.stringify(schema)}</function>`;
      })
      .join('\n');

    return `<functions>\n${functions}\n</functions>`;
  }

  /**
   * 解析 LLM 响应中的工具调用
   * @param {string} response
   * @returns {Array}
   */
  parseFunctionCalls(response) {
    const calls = [];
    
    // 提取 <function_calls> 块
    const functionCallsMatch = response.match(/<function_calls>(.*?)<\/function_calls>/s);
    if (!functionCallsMatch) {
      return calls;
    }

    const functionCallsContent = functionCallsMatch[1];
    
    // 解析多个 <invoke> 调用
    const invokeMatches = functionCallsContent.matchAll(/<invoke name="([^"]+)">(.*?)<\/invoke>/gs);
    
    for (const match of invokeMatches) {
      const toolName = match[1];
      const parametersContent = match[2];
      
      // 解析参数键值对
      const parameters = {};
      const paramMatches = parametersContent.matchAll(/<parameter name="([^"]+)">(.*?)<\/parameter>/gs);
      
      for (const paramMatch of paramMatches) {
        const paramName = paramMatch[1];
        const paramValue = paramMatch[2].trim();
        parameters[paramName] = paramValue;
      }
      
      calls.push({
        tool_name: toolName,
        parameters
      });
    }
    
    return calls;
  }

  /**
   * 批量执行工具调用
   * @param {Array} calls
   * @returns {Array}
   */
  async executeFunctionCalls(calls) {
    const results = [];
    
    for (const call of calls) {
      const { tool_name, parameters } = call;
      const tool = this.getTool(tool_name);
      
      if (!tool) {
        results.push({
          tool_name,
          success: false,
          error: `Tool '${tool_name}' not found`
        });
        continue;
      }
      
      const result = await tool.safeExecute(parameters);
      results.push({
        tool_name,
        ...result
      });
    }
    
    return results;
  }
}

// 全局单例实例
export const toolRegistry = new ToolRegistry(); 